### PR TITLE
Fix leaky subscriptions in the Project Selector

### PR DESCRIPTION
**Pull request by scolsen**
Another step toward finishing #24.

**Note**: This currently contains the quickfix for the homescreen. This change can be ignored and will go away once #72 is merged into issue-24/master.

### CODE REVIEW

**Comment by gino-m** [comment]
> **Note**: This currently contains the quickfix for the homescreen. This change can be ignored and will go away once #72 is merged into issue-24/master.

If you set "base" to the PR with the homescreen change it'll disappear from the diff (click Edit next to PR desc).

File: gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java [hunk 1]
```diff
@@ -15,48 +15,88 @@
...
   ProjectSelectorViewModel(DataRepository dataRepository, AuthenticationManager authManager) {
     this.dataRepository = dataRepository;
     this.projectSummaries = new MutableLiveData<>();
-    this.authManager = authManager;
-  }
+    this.activeProject = new MutableLiveData<>();
+    this.activateProjectErrors = new MutableLiveData<>();
+    this.projectSelections = PublishSubject.create();
+
+    Observable<AuthenticationManager.User> user = authManager.getUser();
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java]: Are we sure it's safe to do all this work in the constructor? Or should it be run in a separate method and invoked from the appropriate lifecycle?

File: gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java [hunk 2]
```diff
@@ -15,48 +15,88 @@
...
     disposeOnClear(
-        authManager
-            .getUser()
-            .flatMap(user -> dataRepository.getProjectSummaries(user))
-            .subscribe(summaries -> projectSummaries.setValue(summaries)));
+        activeProjectStream.subscribe(
+            Result.unwrap(activeProject::setValue, this::onActiveProjectError)));
+    disposeOnClear(
+        availableProjects.subscribe(
+            Result.unwrap(projectSummaries::setValue, this::onProjectSummariesError)));
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java]: Since we're only going half-way to Rx town at this stage, I'm finding it hard to follow what's actually happening now. Let's put our heads together to make these a bit more readable once the fixes are merged.

Also, could you confirm that you've run and tested these changes locally and that the app is WAI?

File: gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java (hunk 1 above)
Comment (scolsen) [reply gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java]: For this particular case, I think it makes sense--the list of available projects needs to be fetched as soon as the VM is instantiated, as these are passed as an original list of options to the Fragment. Thereafter, we wait for selections over that list using the PublishSubject defined here.

Instead of being a stream, I've changed this to refer to only the first user returned by the authManager, to prevent inadvertent actions as other users (not that this is likely).

This also enabled me to be rid of those calls to `Result`.

Open to a better approach if you have one!

**Comment by scolsen** [comment]
> > **Note**: This currently contains the quickfix for the homescreen. This change can be ignored and will go away once #72 is merged into issue-24/master.
> 
> If you set "base" to the PR with the homescreen change it'll disappear from the diff (click Edit next to PR desc).

Good to know, thanks!

File: gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java [hunk 3]
```diff
@@ -15,48 +15,89 @@
...
     this.dataRepository = dataRepository;
     this.projectSummaries = new MutableLiveData<>();
-    this.authManager = authManager;
-  }
+    this.activeProject = new MutableLiveData<>();
+    this.activateProjectErrors = new MutableLiveData<>();
+    this.projectSelections = PublishSubject.create();
+
+    AuthenticationManager.User user =
+        authManager.getUser().blockingFirst(AuthenticationManager.User.ANONYMOUS);
```
Comment (scolsen) [inline gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java]: Are we okay with using [`blockingFirst`](http://reactivex.io/RxJava/javadoc/io/reactivex/Observable.html#blockingFirst-T-)? It's the most concise way, I think, of doing what we want here (getting and keeping a reference only to the first user returned by the user stream, that way the user referenced throughout the interaction is the same).

There are alternatives, but they involve a bit more manipulation since `getUser` returns an Observable.

Comment (gino-m) [reply gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java]: In the future would like to expose a synchronous `currentUser()` method and create the backing Behavior with default anon user, but for the purposes of this fix it looks like the right thing.

File: gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java [hunk 4]
```diff
@@ -15,48 +15,89 @@
...
 
 public class ProjectSelectorViewModel extends AbstractViewModel {
   private static final String TAG = ProjectSelectorViewModel.class.getSimpleName();
 
   private final DataRepository dataRepository;
   private final MutableLiveData<Resource<List<Project>>> projectSummaries;
-  private final AuthenticationManager authManager;
+  private final PublishSubject<Integer> projectSelections;
+  public final MutableLiveData<Project> activeProject;
+  public final MutableLiveData<Throwable> activateProjectErrors;
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java]: I think I need to eat my words about making these public final. Since they're mutable, any client can accidentally update them. We really should expose them as `LiveData` via methods like `projectActivationLiveData()` and `activeProjectLiveData()`. Sorry that I overlooked this!

File: gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java (hunk 3 above)
Comment (scolsen) [reply gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java]: Sounds good! #86!

File: gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java (hunk 4 above)
Comment (scolsen) [reply gnd/src/main/java/com/google/android/gnd/ui/projectselector/ProjectSelectorViewModel.java]: Cool, I've changed these to private and added accessors. I've also updated the fragment appropriately.

I stuck with the existing naming pattern for live data accessors `get<data>()`. We can take on a future effort to rename these with the LiveData suffix, if you want.
